File: userservice/src/main/java/com/example/userservice/user/domain/UserRepository.java
package com.example.userservice.user.domain;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;

import java.util.Optional;

public interface UserRepository extends JpaRepository<Users, Long> {

    @Query("select u from Users u where u.userId = :userId")
    Optional<Users> findByUserId(String userId);

    @Query("select u from Users u where u.email = :email")
    Optional<Users> findByEmail(String email);
}

File: library-app/src/main/kotlin/com/group/libraryapp/domain/user/UserRepository.kt
package com.group.libraryapp.domain.user

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import java.util.*

interface UserRepository : JpaRepository<User, Long> {

    //fun findByName(nameP: String): Optional<User>
    fun findByName(nameP: String): User? //스프링이 알아서 null을 넣어줌.

    @Query("select distinct u from User u left join fetch u.userLoanHistories")
    fun findAllWithHistories(): List<User>
}